training/metrics.py

```diff
@@ -27,59 +25,91 @@ def count_trainable_params(model):
   total_trainable_params = int(sum(np.prod(p.shape.as_list()) for p in object_identity.ObjectIdentitySet(weights)))
   return total_trainable_params
 
-
-def _count_flops_conv2d(layer):
-  if layer.data_format == "channels_first":
-    input_channels = layer.input_shape[1]
-    output_channels, h, w, = layer.output_shape[1:]
-  elif layer.data_format == "channels_last":
-    input_channels = layer.input_shape[3]
-    h, w, output_channels = layer.output_shape[1:]
-  w_h, w_w = layer.kernel_size
-
-  n_mult = h * w * output_channels * input_channels * w_h * w_w
-  n_add = w_h * w_w * input_channels * h * w * output_channels
-
-  flops = n_mult + n_add
+def _linear_layer(layer, N):
+  """
+  Note: This calculates the FLOPs for the unoptimized implementation of any Algebra
+  """
+  input_shape = layer.input_shape
+  output_shape = layer.output_shape
+  if len(input_shape) == 4: # 2D Conv and DepthWise
+    if layer.data_format == "channels_first":
+      input_channels = input_shape[1]
+      output_channels, h, w, = output_shape[1:]
+    elif layer.data_format == "channels_last":
+      input_channels = input_shape[3]
+      h, w, output_channels = output_shape[1:]
+    w_h, w_w = layer.kernel_size
+    if layer.name.lower() == "depthwise_conv2d": 
+      output_channels = 1
+  elif len(input_shape) == 2: # Dense 
+    input_channels = input_shape[1] 
+    output_channels = output_shape[1]
+    w_h, w_w, h, w = 1, 1, 1, 1 # setting this to 1
+  else:
+    raise NotImplementedError("Flops for {layer.name} layer not implemented")
+
+  n_mul = (N**2) * (w_h * w_w * input_channels * output_channels * h * w)
+  n_add = (N*(N-1)) * (w_h * w_w * input_channels * output_channels * h * w)
+  
+  flops = n_mul + n_add
+
+  if N == 1: 
+    flops *= 2 # n_add becomes zero for N = 1
 
   if layer.use_bias:
-    flops += output_channels * h * w
+    flops += output_channels * h * w * N
 
   return int(flops)
 
-
-def _count_flops_relu(layer):
+def _count_flops_relu(layer, N):
   """ Dev note : current tensorflow profiler say ReLU doesn't cost anything...
   """
   # 2 operations per component : compare and assign
-  return reduce(lambda x, y: x*y, layer.output_shape[1:]) * 2
+  return N * (reduce(lambda x, y: x*y, layer.output_shape[1:]) * 2)
 
+def _count_flops_hard_sigmoid(layer, N):
+  return N * (_count_flops_relu(layer, N=1) * 2) # relu + one addtion and one division
 
-def _count_flops_maxpool2d(layer):
-  return layer.pool_size[0] * layer.pool_size[1] * reduce(lambda x, y: x*y, layer.output_shape[1:])
+def _count_flops_hard_swish(layer, N):
+  return N * (_count_flops_hard_sigmoid(layer, N=1) * 2) # hard_sigmoid + 1 multiplication
 
+def _count_flops_maxpool2d(layer, N):
+  return N * (layer.pool_size[0] * layer.pool_size[1] * reduce(lambda x, y: x*y, layer.output_shape[1:]))
```

Comment:
the expression `reduce(lambda x, y: x*y, layer.output_shape[1:]))` is used a lot in this code. Maybe put it in a function with a clear name and explanation ?